From: Yehuda Sadeh <yehuda@redhat.com>
Date: Thu, 30 Jun 2016 17:36:16 -0700
Subject: rgw: forward input data when forwarding set_bucket_version to master

Fixes: http://tracker.ceph.com/issues/16494

Needed to keep input data around to be forwarded correctly. Also, master
does not send any data back, so don't try to parse anything.

Signed-off-by: Yehuda Sadeh <yehuda@redhat.com>
(cherry picked from commit 1eec00bef1e5540bf3e31d1f8fb1645eb64b1e62)

Resolves (with following): rhbz#1350522
---
 src/rgw/rgw_op.cc      | 13 +++++--------
 src/rgw/rgw_op.h       |  1 +
 src/rgw/rgw_rest_s3.cc |  5 +++++
 3 files changed, 11 insertions(+), 8 deletions(-)

diff --git a/src/rgw/rgw_op.cc b/src/rgw/rgw_op.cc
index 4b3b928..76c2428 100644
--- a/src/rgw/rgw_op.cc
+++ b/src/rgw/rgw_op.cc
@@ -1555,20 +1555,17 @@ void RGWSetBucketVersioning::pre_exec()
 
 void RGWSetBucketVersioning::execute()
 {
+  op_ret = get_params();
+  if (op_ret < 0)
+    return;
+
   if (!store->is_meta_master()) {
-    bufferlist in_data;
-    JSONParser jp;
-    op_ret = forward_request_to_master(s, NULL, store, in_data, &jp);
+    op_ret = forward_request_to_master(s, NULL, store, in_data, nullptr);
     if (op_ret < 0) {
       ldout(s->cct, 20) << __func__ << "forward_request_to_master returned ret=" << op_ret << dendl;
     }
     return;
   }
-  
-  op_ret = get_params();
-
-  if (op_ret < 0)
-    return;
 
   if (enable_versioning) {
     s->bucket_info.flags |= BUCKET_VERSIONED;
diff --git a/src/rgw/rgw_op.h b/src/rgw/rgw_op.h
index ecb6dfa..624139b 100644
--- a/src/rgw/rgw_op.h
+++ b/src/rgw/rgw_op.h
@@ -449,6 +449,7 @@ public:
 class RGWSetBucketVersioning : public RGWOp {
 protected:
   bool enable_versioning;
+  bufferlist in_data;
 public:
   RGWSetBucketVersioning() : enable_versioning(false) {}
 
diff --git a/src/rgw/rgw_rest_s3.cc b/src/rgw/rgw_rest_s3.cc
index 49744b9..3cd8f08 100644
--- a/src/rgw/rgw_rest_s3.cc
+++ b/src/rgw/rgw_rest_s3.cc
@@ -746,6 +746,11 @@ int RGWSetBucketVersioning_ObjStore_S3::get_params()
     goto done;
   }
 
+  if (!store->is_meta_master()) {
+    /* only need to keep this data around if we're not meta master */
+    in_data.append(data, len);
+  }
+
   r = parser.get_versioning_status(&enable_versioning);
 
 done:
